### PR TITLE
JUCP: initial commit

### PR DESCRIPTION
**Pull request by netanelyo**

### CODE REVIEW

**Comment by rosenbaumalex** [comment]
nati, 
why do you need the additional 'jucp/' subfolder? 
i think it should be removed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4432/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4434/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
Why do we have a separate license for this ?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4442/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4725/ for details (Mellanox internal link).

File: config/m4/java.m4 [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                            [Compile Java UCX (default is NO).]
+                           )], [with_java=yes], [with_java=no])
+
+
+AS_IF([test "x$with_java" != xno],
+      [AS_IF([test "x$with_java" == xyes],
+             [AS_IF([test -n "$JAVA_HOME"],
+                    [],
+                    [
+                     AC_SUBST([JAVA], [$(readlink -f $(which java))])
```
Comment (miked-mellanox) [inline config/m4/java.m4]: * readlink should be detected, here is an example https://github.com/CZ-NIC/libnetconf/blob/master/configure.in#L317

* s/which/type

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4726/ for details (Mellanox internal link).

File: config/m4/java.m4 [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                            [Compile Java UCX (default is NO).]
+                           )], [with_java=yes], [with_java=no])
+
+
+AS_IF([test "x$with_java" != xno],
+      [AS_IF([test "x$with_java" == xyes],
+             [AS_IF([test -n "$JAVA_HOME"],
+                    [],
+                    [
+                     READLINK_FKNOWN=$(readlink -f xxx 1>/dev/null 2>&1 && echo yes)
```
Comment (yosefe) [inline config/m4/java.m4]: why not AC_CHECK_PROG?

File: config/m4/java.m4 [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+       AS_IF([test -d "$with_java"],
+             [],
+             [AC_MSG_ERROR([Please set JAVA_HOME=<path-to-java>])]
+            )
+
+       save_CPPFLAGS="$CPPFLAGS"
+
+       CPPFLAGS="-I$with_java/include/linux $CPPFLAGS"
+
+        AC_CHECK_HEADER([$with_java/include/linux/jni_md.h],
```
Comment (yosefe) [inline config/m4/java.m4]: need to fix indentation and use less space lines

**Comment by yosefe** [comment]
@shamisp pls take a look

File: config/m4/java.m4 (hunk 2 above)
Comment (shamisp) [reply config/m4/java.m4]: what is -f xxx

**Comment by shamisp** [comment]
Where is the rest of Java code ? Does not make sense to add java configure before we have pending patch for all the rest.

**Comment by shamisp** [comment]
The patch is only 78 code lines. There is enough room here to add all the relevant makefile and a place holder for java headers.

File: config/m4/java.m4 (hunk 2 above)
Comment (netanelyo) [reply config/m4/java.m4]: Sorry I didn't reply, been on vacation.
Changing to AC_CHECK_PROG(..., readlink, ...)

**Comment by netanelyo** [comment]
@shamisp Java doesn't have headers. All the code including the makefiles is ~1100 lines, so I split it to 3 PRs - this one, and 2 PRs of ~500 lines (first one with makefiles and basic Java and C code, second with control path code). I will submit the next PR as soon as this one is approved.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4896/ for details (Mellanox internal link).

**Pull request merged**